src/Bundle.ts

```diff
@@ -12,68 +12,349 @@ import transformBundle from './utils/transformBundle';
-			this.orderedModules.forEach(module => module.includeAllInBundle());
+			safeExportName = variable.exportName = variable.name;
+		}
+
+		let curExport = this.exports[safeExportName];
+		while (curExport) {
+			safeExportName = (variable.exportName || variable.name) + '$' + ++i;
+			curExport = this.exports[safeExportName];
+		}
+
+		curExport = this.exports[safeExportName] = { module, name: undefined, variable };
+		this.exportedVariables.set(variable, safeExportName);
+
+		// if we've just exposed an export of a non-entry-point,
+		// then note we are no longer an entry point bundle
+		// we will then need an entry point facade if this is an entry point module
+		if (this.entryModuleFacade && module.bundle === this && !module.isEntryPoint) {
+			this.entryModuleFacade = false;
+		}
+
+		// if we are reexporting a module in another bundle
+		// then we also have to ensure it is an export there too
+		// and note the name it comes from
+		if (module.bundle !== this && !module.isExternal) {
+			curExport.name = (<Module>module).bundle.ensureExport(module, variable);
 		}
+		else {
+			curExport.name = safeExportName;
+		}
+
+		return safeExportName;
 	}
 
-	setOutputFacade (entryModule: Module) {
-		this.entryModule = entryModule;
+	generateEntryExports (entryModule: Module) {
+		entryModule.getAllExports().forEach(exportName => {
+			const traced = this.traceExport(entryModule, exportName);
+			const variable = traced.module.traceExport(traced.name);
+			this.exports[exportName] = { module: traced.module, name: traced.name, variable };
+			// if we exposed an export in another module ensure it is exported there
+			if (traced.module.bundle !== this && !traced.module.isExternal) {
+				(<Module>traced.module).bundle.ensureExport(traced.module, variable);
+			}
+			this.exportedVariables.set(variable, exportName);
+		});
 	}
 
-	processExternals () {
-		// prune unused external imports
-		this.externalModules = this.graph.externalModules.filter(module => {
-			return module.used || !this.graph.isPureExternalModule(module.id);
+	generateDependencies (entryFacade?: Module) {
+		if (entryFacade) {
+			this.externalModules = [];
+			this.dependencies = [entryFacade.bundle];
+			return;
+		}
+
+		this.externalModules = [];
+		this.dependencies = [];
+
+		this.orderedModules.forEach(module => {
+			module.dependencies.forEach(dep => {
+				if (dep.bundle === this) {
+					return;
+				}
+
+				let depModule: Bundle | ExternalModule;
+				if (dep instanceof Module) {
+					depModule = dep.bundle;
+				} else {
+					// unused pure external modules can be skipped
+					if (!dep.used && this.graph.isPureExternalModule(dep.id)) {
+						return;
+					}
+					depModule = dep;
+				}
+
+				if (!this.dependencies.some(dep => dep === depModule)) {
+					this.dependencies.push(depModule);
+					if (dep.isExternal) {
+						this.externalModules.push(dep);
+					}
+				}
+			});
+		});
+	}
+
+	generateImports () {
```

Comment:
This is a really long homogeneous wall of code that I find really hard to read. Maybe we could at least extract the namespace handling to a separate method, that would cut the size in half. Not sure there are other things that could be extracted.

Reply:
This is effectively a tracing function to populate the final import values. The key thing to follow with a function like this is that it is populating `this.imports` on the chunk object from the module.imports and chunk information. Understand the input and output data structures and the function will make sense.

I've split out a duplicated sub-function here for better readability.

Reply:
Very good, much better than my suggestion 👍